### PR TITLE
enable case-insensitive protocol

### LINKED ISSUE #394: protocols in urls are not case insensitive.
Protocols are case insensitive according to the spec (https://tools.ietf.org/html/rfc3986#section-3.1). It looks like the case insensitivity is handled properly in the get_protocol function above, but most matching of protocols does not go through this function and hence are not actually case insensitive.

### PR DESCRIPTION
**Pull request by guihao-liang**
resolve #394 

protocol in the url remains its original case, which should be handled by the remote fs.

```cpp
dir_archive_impl::archive_index_information read_index_file(std::string index_file) {
  general_ifstream fin(index_file);
...
```

### CODE REVIEW

**Comment by guihao-liang** [comment]
It works as expected.
```
In [6]: tc.SFrame("**HtTpS**://static.turi.com/datasets/bad_csv_example.csv")
Downloading HtTpS://static.turi.com/datasets/bad_csv_example.csv to /var/folders/n9/wscz46nj4_10nz3f2y2r0pnr0000gn/T/turicreate-guihaoliang/39483/e774f2ea-49c5-4b76-b703-5f56b471f70a.csv
Unexpected characters after last column. "a"
Parse failed at token ending at:
        x,y,z,a,^b,c
Successfully parsed 3 tokens:
        0: x
        1: y
        2: z
1 lines failed to parse correctly
Finished parsing file HtTpS://static.turi.com/datasets/bad_csv_example.csv
Parsing completed. Parsed 98 lines in 0.006038 secs.
------------------------------------------------------
Inferred types from first 100 line(s) of file as
column_type_hints=[int,int,int]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------
Unable to interpret "x" as a integer
Parse failed at token ending at:
        x,^y,z,a,b,c
Successfully parsed 0 tokens:
1 lines failed to parse correctly
Finished parsing file HtTpS://static.turi.com/datasets/bad_csv_example.csv
Parsing completed. Parsed 98 lines in 0.005674 secs.
```

**Comment by guihao-liang** [comment]
also works for SFrame, which means SArray should be also working well.

```
In [2]: tc.SFrame("hTTps://static.turi.com/datasets/audioscrobbler")
Downloading hTTps://static.turi.com/datasets/audioscrobbler/dir_archive.ini to /var/folders/n9/wscz46nj4_10nz3f2y2r0pnr0000gn/T/turicreate-guihaoliang/40117/5529cd78-b733-481e-9865-00edede1f3a2.ini
Downloading hTTps://static.turi.com/datasets/audioscrobbler/objects.bin to /var/folders/n9/wscz46nj4_10nz3f2y2r0pnr0000gn/T/turicreate-guihaoliang/40117/cbd7e563-355f-4764-8f09-9f1bd53fe716.bin
Downloading https://static.turi.com/datasets/audioscrobbler/m_74b0dc51.frame_idx to /var/folders/n9/wscz46nj4_10nz3f2y2r0pnr0000gn/T/turicreate-guihaoliang/40117/3db47276-5c3a-4aa3-b8fa-c5464e76c47f.frame_idx
Downloading https://static.turi.com/datasets/audioscrobbler/m_74b0dc51.sidx to /var/folders/n9/wscz46nj4_10nz3f2y2r0pnr0000gn/T/turicreate-guihaoliang/40117/fb0c197c-95fb-402c-ae9f-c3f5bb93f497.sidx
Out[2]: Downloading https://static.turi.com/datasets/audioscrobbler/m_74b0dc51.0000 to /var/folders/n9/wscz46nj4_10nz3f2y2r0pnr0000gn/T/turicreate-guihaoliang/40117/638467e1-13c1-4a78-8f7b-78502cdfb797.0000
Downloading https://static.turi.com/datasets/audioscrobbler/m_74b0dc51.0001 to /var/folders/n9/wscz46nj4_10nz3f2y2r0pnr0000gn/T/turicreate-guihaoliang/40117/54eb228c-b425-467f-a226-a9b921d8a6f5.0001
Downloading https://static.turi.com/datasets/audioscrobbler/m_74b0dc51.0002 to /var/folders/n9/wscz46nj4_10nz3f2y2r0pnr0000gn/T/turicreate-guihaoliang/40117/5e35db24-230c-4b14-b155-21ac935d55d1.0002
Downloading https://static.turi.com/datasets/audioscrobbler/m_74b0dc51.0003 to /var/folders/n9/wscz46nj4_10nz3f2y2r0pnr0000gn/T/turicreate-guihaoliang/40117/1aa5c86f-bc36-4821-8020-7e37efd6285d.0003
Downloading https://static.turi.com/datasets/audioscrobbler/m_74b0dc51.0004 to /var/folders/n9/wscz46nj4_10nz3f2y2r0pnr0000gn/T/turicreate-guihaoliang/40117/b4410bc0-c777-433d-a6f0-7211fda8b6c9.0004
Downloading https://static.turi.com/datasets/audioscrobbler/m_74b0dc51.0005 to /var/folders/n9/wscz46nj4_10nz3f2y2r0pnr0000gn/T/turicreate-guihaoliang/40117/73897063-ba3a-471f-9aa1-b794b44917b3.0005
Downloading https://static.turi.com/datasets/audioscrobbler/m_74b0dc51.0006 to /var/folders/n9/wscz46nj4_10nz3f2y2r0pnr0000gn/T/turicreate-guihaoliang/40117/e07345d4-9305-4a41-ada7-cb9d025ab2a8.0006
Downloading https://static.turi.com/datasets/audioscrobbler/m_74b0dc51.0007 to /var/folders/n9/wscz46nj4_10nz3f2y2r0pnr0000gn/T/turicreate-guihaoliang/40117/12508d44-423b-4d7e-811c-d5037d7f821b.0007
```

**Comment by srikris** [comment]
@guihao-liang is this still WIP?

**Comment by guihao-liang** [comment]
@srikris Yes.

File: src/core/storage/fileio/fs_utils.cpp [hunk 1]
```diff
@@ -440,6 +440,8 @@ EXPORT std::string get_protocol(std::string path) {
   size_t proto = path.find("://");
   if (proto != std::string::npos) {
     std::string ret = boost::algorithm::to_lower_copy(path.substr(0, proto));
+    std::transform(ret.begin(), ret.end(), ret.begin(),
+                   [](unsigned char c) { return std::tolower(c); });
```
Comment (hoytak) [inline src/core/storage/fileio/fs_utils.cpp]: Doesn't the to_lower_copy do the lowercase already?

Comment (guihao-liang) [reply src/core/storage/fileio/fs_utils.cpp]: Ah, let me check. I have un-pushed changes. Let's find the branch the resume my work.

Comment (guihao-liang) [reply src/core/storage/fileio/fs_utils.cpp]: this does everything inline, without an extra copy.

**Comment by guihao-liang** [comment]
`sanitize_s3_url` already lower-cases the protocol name.

File: src/core/storage/fileio/fs_utils.cpp (hunk 1 above)
Comment (hoytak) [reply src/core/storage/fileio/fs_utils.cpp]: Ah, but I'm referring to the `boost::algorithm::to_lower_copy(path.substr(0, proto));` call above.

Comment (hoytak) [reply src/core/storage/fileio/fs_utils.cpp]: https://www.boost.org/doc/libs/1_72_0/doc/html/boost/algorithm/to_lower_copy.html

Comment (guihao-liang) [reply src/core/storage/fileio/fs_utils.cpp]: I see! The name is quite confusing! I will switch to to_lower_copy.

**Comment by guihao-liang** [comment]
Merge this to observe the internal nightly build.

**Comment by guihao-liang** [comment]
passed all relevant internal tests.

**Pull request merged**